pkg/gadget-service/service.go

```diff
@@ -0,0 +1,284 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gadgetservice
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync/atomic"
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	gadgetregistry "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-registry"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	pb "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/local"
+
+	// TODO: Move!
+	_ "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/kubemanager"
+)
+
+type Config struct {
+	SocketFile string
+}
+
+type Service struct {
+	pb.UnimplementedGadgetManagerServer
+	config   *Config
+	listener net.Listener
+	runtime  runtime.Runtime
+	logger   logger.Logger
+	servers  map[*grpc.Server]struct{}
+}
+
+func NewService(defaultLogger logger.Logger) *Service {
+	return &Service{
+		servers: map[*grpc.Server]struct{}{},
+		logger:  defaultLogger,
+	}
+}
+
+func (s *Service) GetInfo(ctx context.Context, request *pb.InfoRequest) (*pb.InfoResponse, error) {
+	catalog, err := s.runtime.GetCatalog()
+	if err != nil {
+		return nil, fmt.Errorf("get catalog: %w", err)
+	}
+
+	catalogJSON, err := json.Marshal(catalog)
+	if err != nil {
+		return nil, fmt.Errorf("marshal catalog: %w", err)
+	}
+	return &pb.InfoResponse{
+		Version: "1.0", // TODO
+		Catalog: catalogJSON,
+	}, nil
+}
+
+func (s *Service) RunGadget(runGadget pb.GadgetManager_RunGadgetServer) error {
+	ctx, cancel := context.WithCancel(runGadget.Context())
+	defer cancel()
+
+	ctrl, err := runGadget.Recv()
+	if err != nil {
+		return err
+	}
+
+	request := ctrl.GetRunRequest()
+	if request == nil {
+		return fmt.Errorf("expected first control message to be gadget request")
+	}
+
+	// Create a new logger that logs to gRPC and falls back to the standard logger when it failed to send the message
+	logger := logger.NewFromGenericLogger(&Logger{
+		send:           runGadget.Send,
+		level:          logger.Level(request.LogLevel),
+		fallbackLogger: s.logger,
+	})
+
+	runtime := s.runtime
+	defer func() {
+		// Try to send a done message
+		runGadget.Send(&pb.GadgetEvent{Type: pb.EventTypeGadgetDone})
+	}()
+
+	gadgetDesc := gadgetregistry.Get(request.GadgetCategory, request.GadgetName)
+	if gadgetDesc == nil {
+		logger.Errorf("gadget not found: %s/%s", request.GadgetCategory, request.GadgetName)
+		return nil
```

Comment:
I'll have to look into it again a bit more - it's not really clear to me where that error message would end up right now.

Reply:
> Yeah, the error should be returned as it is a critical error because it causes the gadget not to run.

I changed it to what you suggested here. IMHO, it's much cleaner now.

> BTW, I think this kind of error should be the actual error returned by the cobra command

This should now be the case.